flocker/ca/functional/test_validation.py

```diff
@@ -0,0 +1,274 @@
+    """
+    bad_name, another_bad_name = {"user", "node", "control"}.difference(
+        {good_certificate_name})
+
+    class ValidationTests(TestCase):
+        """
+        Tests to ensure correct validation of a specific type of certificate.
+
+        :ivar CertificateSet good_ca: The certificates for the CA we expect.
+
+        :ivar CertificateSet another_ca: A different CA's certificates.
+        """
+        def setUp(self):
+            self.good_ca, self.another_ca = get_sets()
+
+        def assert_validates(self, credential):
+            """
+            Asserts that a TLS handshake is successfully established between a
+            client using the validation logic and a server based on the
+            given credential.
+
+            :param credential: The high-level credential to use.
+
+            :return ``Deferred``: Fires on success.
+            """
+            credential = credential.credential
+            private_certificate = PrivateCertificate.fromCertificateAndKeyPair(
+                credential.certificate, credential.keypair.keypair)
+            peer_context_factory = private_certificate.options()
+
+            port = find_free_port()[1]
+            validating_context_factory = context_factory_fixture(
+                port, self.good_ca)
+
+            if validator_is_client:
+                client_context_factory = validating_context_factory
+                server_context_factory = peer_context_factory
+            else:
+                server_context_factory = validating_context_factory
+                client_context_factory = peer_context_factory
+
+            server_endpoint = SSL4ServerEndpoint(reactor, port,
+                                                 server_context_factory,
+                                                 interface='127.0.0.1')
+            d = server_endpoint.listen(
+                ServerFactory.forProtocol(SendingProtocol))
+            d.addCallback(lambda port: self.addCleanup(port.stopListening))
+            validating_endpoint = SSL4ClientEndpoint(
+                reactor, b"127.0.0.1", port, client_context_factory)
+            client_protocol = ReceivingProtocol()
+            result = connectProtocol(validating_endpoint, client_protocol)
+            result.addCallback(lambda _: client_protocol.result)
+            return result
+
+        def assert_does_not_validate(self, credential):
+            """
+            Asserts that a TLS handshake fails to happen between a client using
+            the validation logic and a server based on the given
+            credential.
+
+            :param FlockerCredential credential: The private key/certificate
+                to use for the server.
+
+            :return ``Deferred``: Fires on success (i.e. if no TLS handshake is
+                established).
+            """
+            return self.assertFailure(self.assert_validates(credential),
+                                      SSLError)
```

Comment:
It would probably be somewhat good to try to make an assertion about what the error was that leads to the `SSLError`.  I admit that might be difficult given the form in which OpenSSL/pyOpenSSL present failures.  The security-sensitive nature of this code probably merits an attempt, though, so we can be sure that when one of these negative-case tests passes it's because of validation and not because some other stupid bugs.